### DISCUSSION TITLE
Tip: Use screencasts to debug eeschema/pcbnew timeouts

### DISCUSSION THREAD

**Discussion by festlv** [INTI-CMNB/KiBot#231] [General]
I initially wanted to start this discussion on how to debug timeout of eeschema/pcbnew. 

However, when trying to create a minimal reproducible sample, I noticed that there already is a feature to debug these kinds of issues: a screen recording of interacting with eeschema in the hidden X11 server is being recorded and is available in the Schematic folder (export_eeschema_screencast.ogv).

Thank you @set-soft, this helped to solve my issue (dockerized kicad trying to write to ~/.local which is failing as it's not mounted)- I am leaving this discussion here as it may help others.

**Comment by festlv** [comment]
Apparently, the screencasts are only recorded/saved when --verbose flag is passed to kibot, right?

**Reply by set-soft** [reply]
Hi @festlv !

Yes, the recording is enabled only when you are debugging. The OGV compression can eat one CPU core and, in some cases, generate a 10 MiB file.

Also check the KiAuto options, if you need to debug some complex stuff you can take a look at KiBot debug output. It will say how KiAuto is called and you can use it directly. In this case you can use more specific stuff. The options are explained [here](https://github.com/INTI-CMNB/KiAuto#common-options)

Connecting to the virtual X11 display you can also interact with it.

**Comment by cdwilson** [comment]
I noticed these files earlier this morning when I enabled the verbose flag for the first time.  However, all the screencasts in my outputs seem to be 0:00 in length...  @festlv how were you able to generate the full recording?

Here's an example:

[bom_xml_eeschema_screencast.ogv.zip](https://github.com/INTI-CMNB/KiBot/files/9162638/bom_xml_eeschema_screencast.ogv.zip)

**Comment by festlv** [comment]
@cdwilson: how are you running kibot? My first guesses would be that you're either running a different version of kibot, or perhaps you're missing some dependencies if you're running kibot locally?

I'm running kibot inside docker as follows (in the same directory as project):

```
WORKDIR=.
SUBDIR=.
docker run --rm -it \
    -v $(pwd)/$WORKDIR:/workdir \
    --workdir="/workdir" \
    --volume="/usr/share/kicad/3dmodels:/3dmodels:ro" \
    --env KICAD6_3DMODEL_DIR=/3dmodels \
    --env KIPRJMOD=/workdir/$SUBDIR \
    setsoft/kicad_auto_test:ki6 /bin/bash -c "cd /workdir/$SUBDIR; kibot -v"
````

**Reply by cdwilson** [reply]
I'm also running KiBot in docker (on a Mac, but it shouldn't make any difference).  the screencast files are being created, but they don't seem to be doing the full recording (they are 0:00 length).

**Reply by set-soft** [reply]
Hi @cdwilson !
I tried the .zip you uploaded, it contains a truncated video, but no 0 length.
I always decompose the video into individual frames, is much more easy to analyze.
I use `ffmpeg` tool, like this:
```
ffmpeg -i *.ogv  -vsync 0 out%03d.png
```
In the case of your video the last frame is:
![out005](https://user-images.githubusercontent.com/22823627/180410846-f86b786e-130c-4cea-a428-c7b1757ad205.png)

You could also try to add a call to `sync` after running KiBot.

**Reply by cdwilson** [reply]
Thanks @set-soft.  I had previously tried to open the video in VLC and since the video is only 5 frames, it played so fast that I didn't see anything and assumed it was 0:00 length... 🤦🏻‍♂️

> I always decompose the video into individual frames, is much more easy to analyze.
> I use ffmpeg tool, like this:
> 
> ```
> ffmpeg -i *.ogv  -vsync 0 out%03d.png
> ```

This worked for me, but just FYI, it looks like there is [an open bug](https://github.com/Enselic/recordmydesktop/issues/23) in recordmydesktop that causes the following error when I run the command above:

```
[ogg @ 0x400009de40] Broken file, keyframe not correctly marked.
```

From [a comment in the issue](https://github.com/Enselic/recordmydesktop/issues/23#issuecomment-1147327560), it looks like this can be fixed by running the `.ogv` through ffmpeg:

```
ffmpeg -i original.ogv -c copy cleaned.ogv
```

Do you know if there a way to change the capture frame rate for the screencast? (i.e. capture more frames/second?)

**Reply by set-soft** [reply]
Hi @cdwilson !
Why do you need more FPSs? Note that you get 5 frames not because of time, but because the recorder found 5 different frames.

**Reply by cdwilson** [reply]
> Note that you get 5 frames not because of time, but because the recorder found 5 different frames.

@set-soft oh, I see.  I think I had the wrong expectations about what the screencast recording would look like.  I asked the question about FPS because I (very naively) assumed that increasing the FPS of the recording would generate a smoother screencast with more details, but it sounds like that's not the way things are working (I have no idea how any of the virtual X11 display works in KiBot).  Are you saying that the virtual X11 display is only generating 5 frames to begin with, so the recording is only capturing those 5 frames?

**Reply by set-soft** [reply]
@cdwilson I can't say for sure they are just 5, but they are closer to it. X servers can join drawing requests and perform them after a flush, so you see these changes together, not separated.
What are you trying to catch?

**Reply by cdwilson** [reply]
@set-soft thanks for the clarifications.  I'm not trying to catch anything specific right now.  I just tried this out for the first time and I wanted to make sure that I'm getting the same screencast results as everyone else when enabling the verbose options.  It sounds like what I'm getting is the expected results.  Sorry for all the noise on this!
